### PR TITLE
Allowing injecting in constructor

### LINKED ISSUE #26: Let's Talk: Constructor Injection
The new 1.15 snapshots move almost all of `MinecraftClient`'s setup stuff into the constructor. This is a *massive* hassle for any mod that wants to add its own resource pack stuff, which includes Fabric API. Fabric API has [resorted to a `@Redirect`](https://github.com/FabricMC/fabric/blob/1.15/fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/mixin/resource/loader/MixinMinecraftGame.java#L74), but this means that nobody else can use that point, since `@Redirect` is collidable. The reason we have this problem is because `@Inject` only allows injection to the `RETURN` of constructors. Do we want to open that up to any other injection points? The main worry is preventing anything from being put before a `super` call.

### PR DESCRIPTION
**Pull request by liach**
It is bad to assume that everyone does business logic outside of constructor. We don't want redirect conflict hell

Fixes #26

### CODE REVIEW

**Comment by Juuxel** [comment]
Doesn't this allow cancelling final field initialization?

**Comment by liach** [comment]
Will do a few tweaks to prevent `cancellable` then, I guess

**Comment by kvverti** [comment]
How about `static` injections before the super/this delegation?

**Comment by liach** [comment]
You cannot call static methods before super/this delegation in constructors unless you are calling it for parameters. Imo in that case redirects are better. Redirect, Modify arg, Modify variable is already `ALLOW_ALL` when passed into checking.

**Comment by liach** [comment]
This patch works.
image
Mixin code:
```java
@Mixin(MinecraftClient.class)
public abstract class ClientMixin {
    @Inject(method = "<init>", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/MinecraftClient;initializeSearchableContainers()V"))
    private void injectInit(RunArgs args, CallbackInfo ci) {
        System.out.println("HELLO I AM INITIALIZING SEARCHABLE CONTAINERS");
    }
}
```
---
Also the `cancellable = true` check works as well:
image
```java
@Mixin(MinecraftClient.class)
public abstract class ClientMixin {
    @Inject(method = "<init>", cancellable = true,
            at = @At(value = "INVOKE", target = "Lnet/minecraft/client/MinecraftClient;initializeSearchableContainers()V"))
    private void injectInit(RunArgs args, CallbackInfo ci) {
        System.out.println("HELLO I AM INITIALIZING SEARCHABLE CONTAINERS");
    }
}
```

**Comment by i509VCB** [comment]
Hmm, what's does the output look like (Dmixin.debug.output=true)

**Comment by liach** [comment]
@i509VCB `javap` result of `MinecraftClient` [here](https://gist.github.com/0f8449e07387ff7486d2babd0a252328) (Note: Fabric API patches exist as well)

**Comment by liach** [comment]
@sfPlayer1 the before super injection does properly fail  i.e. `HEAD` 
image

File: build.gradle [hunk 1]
```diff
@@ -23,7 +23,7 @@ group = 'net.fabricmc'
 archivesBaseName = 'sponge-mixin'
 version = buildVersion
 
-ext.buildNumber = '0'
+ext.buildNumber = '571428'
```
Comment (liach) [inline build.gradle:26]: hmm, i should revert this for now

**Comment by Runemoro** [comment]
Disallowing cancellation is unnecessary. There's nothing wrong with final fields being left uninitialized. Here's some vanilla Java code that accesses a final field before it's initialized:

```java
public class Test {
    private final int a;

    public Test(int a) {
        this.a = f();
        System.out.println(a); // prints 1
    }

    private int f() {
        System.out.println(a); // a is accessed before being assigned, prints 0
        return 2;
    }
}
```

**Comment by TehcJS** [comment]
pls merge w/ master tanqs


i need feature :(

**Review by YTG1234 — APPROVED** [review]

**Comment by modmuss50** [comment]
Going to test and merge this in via: https://github.com/FabricMC/Mixin/pull/40

**Pull request closed without merging**